nvflare/private/fed/server/fed_server.py

```diff
@@ -634,9 +634,24 @@ def Heartbeat(self, request, context):
             if self.admin_server:
                 self.admin_server.client_heartbeat(token)
 
+            abort_runs = self._sync_client_jobs(request)
             summary_info = fed_msg.FederatedSummary()
+            if abort_runs:
+                del summary_info.abort_jobs[:]
+                summary_info.abort_jobs.extend(abort_runs)
+                display_runs = ",".join(abort_runs)
+                self.logger.info(
+                    f"These jobs: {display_runs} are not running on the server. "
+                    f"Ask client: {client_name} to abort these runs."
+                )
             return summary_info
 
+    def _sync_client_jobs(self, request):
+        client_jobs = request.jobs
+        server_jobs = self.engine.run_processes.keys()
```

Comment:
This is not quite right, since engine.run_processes return ALL processes. What's needed here are the processes corresponding to jobs running on this particular client! There could be cases (though extremely rare) that the client is running job A and the server is also running job A but for different clients. In this case the job A on the client still should be aborted.

The server already keeps jobs under each client (I think), and server_jobs here should be the jobs for this client.